# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [1]:
#load python packages
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline

os.getcwd()

'/Users/student/Desktop/SpringBoardDSCareerTrack/Springboard/Guided Capstone Ski'

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [2]:
path="/Users/student/Desktop/SpringBoardDSCareerTrack/Springboard/Guided Capstone Ski/data"
os.chdir(path)
os.getcwd()

ski_data = pd.read_csv('Ski_data_Exploratory.csv')
ski_data.head(5)

Name    state  summit_elev  vertical_drop  trams  fastEight  \
0       Alyeska Resort   Alaska         3939           2500      1        0.0   
1  Eaglecrest Ski Area   Alaska         2600           1540      0        0.0   
2     Hilltop Ski Area   Alaska         2090            294      0        0.0   
3     Arizona Snowbowl  Arizona        11500           2300      0        0.0   
4  Sunrise Park Resort  Arizona        11100           1800      0        0.0   

   fastSixes  fastQuads  quad  triple  ...  SkiableTerrain_ac  Snow Making_ac  \
0          0          2     2       0  ...             1610.0           113.0   
1          0          0     0       0  ...              640.0            60.0   
2          0          0     0       1  ...               30.0            30.0   
3          1          0     2       2  ...              777.0           104.0   
4          0          1     2       3  ...              800.0            80.0   

   daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  AdultWeekend  \
0             150.0       60.0            669.0          65.0          85.0   
1              45.0       44.0            350.0          47.0          53.0   
2             150.0       36.0             69.0          30.0          34.0   
3             122.0       81.0            260.0          89.0          89.0   
4             115.0       49.0            250.0          74.0          78.0   

   projectedDaysOpen  NightSkiing_ac  clusters  
0              150.0           550.0         0  
1               90.0             0.0         0  
2              152.0            30.0         0  
3              122.0             0.0         1  
4              104.0            80.0         1  

[5 rows x 26 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [3]:
state_dummy = ski_data['state']
ski_data = pd.concat([ski_data.drop(['state'], axis=1), pd.get_dummies(state_dummy)], axis=1)
ski_data

Name  summit_elev  vertical_drop  trams  \
0                       Alyeska Resort         3939           2500      1   
1                  Eaglecrest Ski Area         2600           1540      0   
2                     Hilltop Ski Area         2090            294      0   
3                     Arizona Snowbowl        11500           2300      0   
4                  Sunrise Park Resort        11100           1800      0   
..                                 ...          ...            ...    ...   
325               Meadowlark Ski Lodge         9500           1000      0   
326          Sleeping Giant Ski Resort         7428            810      0   
327                   Snow King Resort         7808           1571      0   
328  Snowy Range Ski & Recreation Area         9663            990      0   
329                White Pine Ski Area         9500           1100      0   

     fastEight  fastSixes  fastQuads  quad  triple  double  ...  Rhode Island  \
0          0.0          0          2     2       0       0  ...             0   
1          0.0          0          0     0       0       4  ...             0   
2          0.0          0          0     0       1       0  ...             0   
3          0.0          1          0     2       2       1  ...             0   
4          0.0          0          1     2       3       1  ...             0   
..         ...        ...        ...   ...     ...     ...  ...           ...   
325        0.0          0          0     0       1       1  ...             0   
326        0.0          0          0     0       1       1  ...             0   
327        0.0          0          0     1       1       1  ...             0   
328        0.0          0          0     0       1       3  ...             0   
329        0.0          0          0     0       2       0  ...             0   

     South Dakota  Tennessee  Utah  Vermont  Virginia  Washington  \
0               0          0     0        0         0           0   
1               0          0     0        0         0           0   
2               0          0     0        0         0           0   
3               0          0     0        0         0           0   
4               0          0     0        0         0           0   
..            ...        ...   ...      ...       ...         ...   
325             0          0     0        0         0           0   
326             0          0     0        0         0           0   
327             0          0     0        0         0           0   
328             0          0     0        0         0           0   
329             0          0     0        0         0           0   

     West Virginia  Wisconsin  Wyoming  
0                0          0        0  
1                0          0        0  
2                0          0        0  
3                0          0        0  
4                0          0        0  
..             ...        ...      ...  
325              0          0        1  
326              0          0        1  
327              0          0        1  
328              0          0        1  
329              0          0        1  

[330 rows x 60 columns]

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [4]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = ski_data.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = ski_data['AdultWeekend'] 

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [5]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [6]:
# all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [7]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)

## Review Model Outcomes — Iterate over additional models as needed

In [8]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model

import sklearn.metrics 
sklearn.metrics.explained_variance_score(y_test, y_pred)

0.938713414488971

In [9]:
sklearn.metrics.mean_absolute_error(y_test, y_pred)

4.984189666762277

**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [10]:
lm.intercept_

64.12083081791931

**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [11]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])

Coefficient
summit_elev        1.901170e+00
vertical_drop      1.049355e+00
trams              3.116284e+12
fastEight          3.063613e+11
fastSixes          3.626843e+12
fastQuads          1.223423e+13
quad               7.303073e+12
triple             9.010989e+12
double             1.010123e+13
surface            1.146255e+13
total_chairs       3.227158e+13
Runs               1.581421e+00
TerrainParks       2.248535e-01
LongestRun_mi      5.522461e-01
SkiableTerrain_ac  1.421661e-01
Snow Making_ac     9.035645e-01
daysOpenLastYear   9.008789e-01
yearsOpen          4.594727e-01
averageSnowfall    1.306152e-01
AdultWeekday       1.931641e+01
projectedDaysOpen  7.958984e-02
NightSkiing_ac     6.987305e-01
clusters           6.235352e-01
Alaska             3.582656e+12
Arizona            2.929696e+12
California         9.214240e+12
Colorado           9.415802e+12
Connecticut        4.611023e+12
Idaho              7.066019e+12
Illinois           4.130565e+12
Indiana            2.929696e+12
Iowa               3.582656e+12
Maine              6.148149e+12
Maryland           2.074763e+12
Massachusetts      6.775827e+12
Michigan           1.068693e+13
Minnesota          7.608135e+12
Missouri           2.929696e+12
Montana            7.066019e+12
Nevada             4.130565e+12
New Hampshire      8.107659e+12
New Jersey         2.929696e+12
New Mexico         6.148149e+12
New York           1.132416e+13
North Carolina     5.043346e+12
Ohio               4.611023e+12
Oregon             6.470616e+12
Pennsylvania       8.792809e+12
Rhode Island       2.074763e+12
South Dakota       2.929696e+12
Tennessee          2.074763e+12
Utah               7.342973e+12
Vermont            7.862697e+12
Virginia           4.130565e+12
Washington         6.470616e+12
West Virginia      4.130565e+12
Wisconsin          8.107659e+12
Wyoming            5.805553e+12

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [12]:
# Import our csv data again into ski_data_2
ski_data_2 = pd.read_csv('Ski_data_Exploratory.csv')
ski_data_2.head(5)

# Drop the state column, name and adult weekend
X = ski_data_2.drop(['state','Name','AdultWeekend'], axis=1)
y = ski_data_2['AdultWeekend']
scaler = preprocessing.StandardScaler().fit(X)
X_scaled=scaler.transform(X)
y = y.ravel()
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

In [13]:
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

y_pred = model.predict(X_test)

print('Explained Variance: ', sklearn.metrics.explained_variance_score(y_test, y_pred))
print('Mean Absolute Error: ', sklearn.metrics.mean_absolute_error(y_test, y_pred))
print('Intercept: ', lm.intercept_)

pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])

Explained Variance:  0.9332096763924219
Mean Absolute Error:  5.177535976985603
Intercept:  64.09065653180949


Coefficient
summit_elev           4.715922
vertical_drop         1.081477
trams                 0.147532
fastEight             0.251301
fastSixes             0.302298
fastQuads             1.029596
quad                  1.021596
triple                1.342061
double                0.166828
surface               1.391288
total_chairs          0.576404
Runs                  2.007674
TerrainParks          0.057474
LongestRun_mi         0.586427
SkiableTerrain_ac     0.429175
Snow Making_ac        0.710863
daysOpenLastYear      0.440934
yearsOpen             0.313778
averageSnowfall       1.088039
AdultWeekday         19.839512
projectedDaysOpen     0.077105
NightSkiing_ac        0.713011
clusters              6.025212

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [14]:
# Import our csv data again into ski_data_2
ski_data_3 = pd.read_csv('Ski_data_Exploratory.csv')
ski_data_3.head(5)

# Drop the state column, name and adult weekend
X = ski_data_3.drop(['state', 'summit_elev', 'Name', 'AdultWeekend'], axis=1)   # base elev was dropped in step 3: Exploratory analysis
y = ski_data_3['AdultWeekend']
scaler = preprocessing.StandardScaler().fit(X)
X_scaled=scaler.transform(X)
y = y.ravel()
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

In [15]:
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

y_pred = model.predict(X_test)

print('Explained Variance: ',sklearn.metrics.explained_variance_score(y_test, y_pred))
print('Mean Absolute Error: ', sklearn.metrics.mean_absolute_error(y_test, y_pred))
print('Intercept: ', lm.intercept_)

Explained Variance:  0.9325709717917094
Mean Absolute Error:  5.1971906410193895
Intercept:  64.07087665186458


In [16]:
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])

Coefficient
vertical_drop         2.156824
trams                 0.335731
fastEight             0.093043
fastSixes             0.417656
fastQuads             0.778229
quad                  1.097195
triple                1.319183
double                0.185192
surface               1.156177
total_chairs          0.574376
Runs                  1.488034
TerrainParks          0.238219
LongestRun_mi         0.512880
SkiableTerrain_ac     0.134879
Snow Making_ac        0.810057
daysOpenLastYear      0.662052
yearsOpen             0.296418
averageSnowfall       0.905639
AdultWeekday         20.080769
projectedDaysOpen     0.001563
NightSkiing_ac        0.389299
clusters              2.642409

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | 0.939 | 4.984 |-|
| Model 2. | 0.933 | 5.178 |'state'|
| Model 3. | 0.933 | 5.197 |'state','summit_elev','base_elev'|

Model Selection:

It seems that while Model 1 and Model 2 are more accurate, Model 3 proves most insightful for features that drive ski resort lift ticket prices as they have features that management can control whereas the driving features for Model 1 is mostly the states, which we can't control.  